spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,516 @@
+      case None => (intervalGetTableVersion, true)
+    }
+    if (version < 0) {
+      return None
+    }
+
+    getStartingOffsetFromSpecificDeltaVersion(version, isStartingVersion, limits)
+  }
+
+  override def getDefaultReadLimit: ReadLimit = {
+    new AdmissionLimits().toReadLimit
+  }
+
+  override def latestOffset(startOffset: streaming.Offset, limit: ReadLimit): streaming.Offset = {
+    val limits = AdmissionLimits(limit)
+
+    val currentOffset = if (previousOffset == null) {
+      getStartingOffset(limits)
+    } else {
+      getNextOffsetFromPreviousOffset(previousOffset, limits)
+    }
+    logDebug(s"previousOffset -> currentOffset: $previousOffset -> $currentOffset")
+    currentOffset.orNull
+  }
+
+  override def getOffset: Option[Offset] = {
+    throw new UnsupportedOperationException(
+      "latestOffset(Offset, ReadLimit) should be called instead of this method")
+  }
+
+  override def getBatch(startOffsetOption: Option[Offset], end: Offset): DataFrame = {
+    val endOffset = DeltaSharingSourceOffset(tableId, end)
+    previousOffset = endOffset // Proceed the offset, and for recovery,
+
+    val (startVersion,
+    startIndex,
+    isStartingVersion,
+    startSourceVersion) = if (startOffsetOption.isEmpty) {
+      getStartingVersion match {
+        case Some(v) =>
+          // startingVersion is provided by the user
+          (v, -1L, false, None)
+
+        case _ =>
+          // startingVersion is NOT provided by the user
+          if (endOffset.isStartingVersion) {
+            // get all files in this version if endOffset is startingVersion
+            (endOffset.tableVersion, -1L, true, None)
+          } else {
+            assert(
+              endOffset.tableVersion > 0, s"invalid tableVersion in endOffset: $endOffset")
+            // Load from snapshot `endOffset.tableVersion - 1L` if endOffset is not
+            // startingVersion
+            (endOffset.tableVersion - 1L, -1L, true, None)
+          }
+      }
+    } else {
+      val startOffset = DeltaSharingSourceOffset(tableId, startOffsetOption.get)
+      if (startOffset == endOffset) {
+        // This happens only if we recover from a failure and `MicroBatchExecution` tries to call
+        // us with the previous offsets. The returned DataFrame will be dropped immediately, so we
+        // can return any DataFrame.
+        return DeltaSharingScanUtils.internalCreateDataFrame(spark, schema)
+      }
+      (startOffset.tableVersion, startOffset.index, startOffset.isStartingVersion,
+        Some(startOffset.sourceVersion))
+    }
+    logDebug(s"start: $startOffsetOption end: $end")
+    val createdDf = getFileChangesAndCreateDataFrame(
+      startVersion, startIndex, isStartingVersion, endOffset
+    )
+
+    createdDf
+  }
+
+  override def stop(): Unit = {}
+
+  override def toString(): String = s"DeltaSharingSource[${deltaLog.table.toString}]"
+
+  trait DeltaSharingSourceAdmissionBase { self: AdmissionLimits =>
```

Comment:
Question: Why do we use this design pattern with a trait and a self directive?
Why not simply move this admit function into AdmissionsLimits itself?